File: java/data_structures/graph/AbstractGraph.java
package data_structures.graph;

import java.util.Objects;

/**
 * Base abstraction for Graph structures.
 *
 * Defines the minimal contract that all graph implementations
 * (adjacency list, adjacency matrix, directed, undirected)
 * must satisfy.
 */
public abstract class AbstractGraph<V> {

    private final boolean directed;

    protected AbstractGraph(boolean directed) {
        this.directed = directed;
    }

    public boolean isDirected() {
        return directed;
    }

    // ---------------------------------
    // Abstract API (must be implemented)
    // ---------------------------------

    public abstract void addVertex(V v);

    public abstract void addEdge(V u, V v);

    public abstract Iterable<V> neighbors(V v);

    public abstract boolean hasEdge(V u, V v);

    public abstract Iterable<V> vertices();

    // ---------------------------------
    // Concrete shared behavior
    // ---------------------------------

    public boolean containsVertex(V v) {
        for (V x : vertices()) {
            if (Objects.equals(x, v)) return true;
        }
        return false;
    }

    public int size() {
        int count = 0;
        for (V ignored : vertices()) count++;
        return count;
    }

    /** Degree of v. For directed graphs, this is the out-degree. */
    public int degree(V v) {
        int count = 0;
        for (V ignored : neighbors(v)) count++;
        return count;
    }

    public boolean isEmpty() {
        return size() == 0;
    }
}

File: csharp/data_structure/graph/AbstractGraph.cs
using System;
using System.Collections.Generic;

namespace DataStructures.Graph
{
    public abstract class AbstractGraph<T>
        where T : notnull
    {
        protected AbstractGraph(bool directed = false)
        {
            Directed = directed;
        }

        public bool Directed { get; }

        // ---------------------------------
        // Abstract API (must be implemented)
        // ---------------------------------

        public abstract void AddVertex(T v);

        public abstract void AddEdge(T u, T v);

        public abstract IEnumerable<T> Neighbors(T v);

        public abstract bool HasEdge(T u, T v);

        public abstract IEnumerable<T> Vertices();

        // ---------------------------------
        // Concrete shared behavior
        // ---------------------------------

        public bool ContainsVertex(T v)
        {
            foreach (var x in Vertices())
            {
                if (EqualityComparer<T>.Default.Equals(x, v))
                    return true;
            }
            return false;
        }

        public int Size()
        {
            int count = 0;
            foreach (var _ in Vertices()) count++;
            return count;
        }

        /// <summary>
        /// Degree of v. For directed graphs, this is the out-degree.
        /// </summary>
        public int Degree(T v)
        {
            int count = 0;
            foreach (var _ in Neighbors(v)) count++;
            return count;
        }

        public bool IsEmpty() => Size() == 0;
    }
}
